commit 35643762a23097b7e97cffec68a2b5a91cefa784
Author: P33M <p33m@github.com>
Date:   Thu Feb 15 11:22:44 2018 +0000

    dwc_otg: add smp_mb() to prevent driver state corruption on boot
    
    Occasional crashes have been seen where the FIQ code dereferences
    invalid/random pointers immediately after being set up, leading to
    panic on boot.
    
    The crash occurs as the FIQ code races against hcd_init_fiq() and
    the hcd_init_fiq() code races against the outstanding memory stores
    from dwc_otg_hcd_init(). Use explicit barriers after touching
    driver state.

diff --git a/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c b/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
index 992269d61ecf..2ccd2a80dbef 100644
--- a/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
+++ b/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
@@ -519,6 +519,11 @@ static void hcd_init_fiq(void *cookie)
 		DWC_ERROR("Can't get FIQ irq");
 		return;
 	}
+	/*
+	 * We could take an interrupt immediately after enabling the FIQ.
+	 * Ensure coherency of hcd->fiq_state.
+	 */
+	smp_mb();
 	enable_fiq(irq);
 	local_fiq_enable();
 #endif
@@ -598,7 +603,11 @@ int hcd_init(dwc_bus_dev_t *_dev)
 
 	if (fiq_enable) {
 		if (num_online_cpus() > 1) {
-			/* bcm2709: can run the FIQ on a separate core to IRQs */
+			/*
+			 * bcm2709: can run the FIQ on a separate core to IRQs.
+			 * Ensure driver state is visible to other cores before setting up the FIQ.
+			 */
+			smp_mb();
 			smp_call_function_single(1, hcd_init_fiq, otg_dev, 1);
 		} else {
 			smp_call_function_single(0, hcd_init_fiq, otg_dev, 1);
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1136_35643762a23097b7e97cffec68a2b5a91cefa784.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1136_35643762a23097b7e97cffec68a2b5a91cefa784.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
